ident/map.go

```diff
@@ -0,0 +1,256 @@
+// ```
+// VALUE_TYPE=YourValueType go generate
+// ```
+// This will output to stdout the generated source file to use for your map.
+// It uses linear probing by incrementing the number of the hash created when
+// hashing the identifier if there is a collision. The hashing algorithm used is
+// fnv hash.
+// The map is a value type and not an interface to allow for less painful
+// upgrades when adding/removing methods, it is not likely to need mocking so
+// an interface would not be super useful either.
+type Map struct {
+	// lookup uses fnv hash of the identifier for the key and the value
+	// wraps the value type and the ident (used to ensure lookup is correct
+	// when dealing with collisions), we use fnv partially because lookups of
+	// maps with uint64 keys has a fast path for Go.
+	lookup map[MapHash]MapEntry
+
+	hash func(ID) MapHash
+
+	opts MapOptions
+}
+
+// MapHash is the hash for a given map entry, this is public to support
+// iterating over the map using a native Go for loop.
+type MapHash uint64
+
+// MapEntry is an entry in the map, this is public to support iterating
+// over the map using a native Go for loop.
+type MapEntry struct {
+	// key is used to check equality on lookups to resolve collisions
+	key mapKey
+	// value type stored
+	value ValueType
+}
+
+type mapKey struct {
+	id       ID
+	finalize bool
+}
+
+// Key returns the map entry key.
+func (e MapEntry) Key() ID {
+	return e.key.id
+}
+
+// Value returns the map entry value.
+func (e MapEntry) Value() ValueType {
+	return e.value
+}
+
+// MapOptions is a set of options used when creating an identifier map.
+type MapOptions struct {
+	// Size is the initial size of the map to create.
+	Size int64
+	// Pool is an underlying pool to use when taking copies of the ID
+	// when using the default Set semantics.
+	Pool Pool
+}
+
+// NewMap returns a new identifier map.
+func NewMap(opts MapOptions) *Map {
+	var lookup map[MapHash]MapEntry
+	if opts.Size > 0 {
+		// Only use the constructor with specific size if set to let
+		// go's default size be used if no specific size was set
+		lookup = make(map[MapHash]MapEntry, opts.Size)
+	} else {
+		lookup = make(map[MapHash]MapEntry)
+	}
+	return &Map{
+		lookup: lookup,
+		hash:   hashFnv,
+		opts:   opts,
+	}
+}
+
+func (m *Map) copyKey(key ID) ID {
+	if m.opts.Pool == nil {
+		bytes := append([]byte(nil), key.Data().Get()...)
+		return BinaryID(checked.NewBytes(bytes, nil))
+	}
+	return m.opts.Pool.Clone(key)
+}
+
+// Get returns a value in the map for an identifier if found.
+func (m *Map) Get(id ID) (ValueType, bool) {
+	hash := m.hash(id)
+	for v, ok := m.lookup[hash]; ok; v, ok = m.lookup[hash] {
+		if v.key.id.Equal(id) {
+			return v.value, true
+		}
+		// Linear probe to "next" to this entry (really a rehash)
+		hash++
+	}
+	var empty ValueType
+	return empty, false
+}
+
+// Set will set the value for an identifier.
+func (m *Map) Set(id ID, value ValueType) {
```

Comment:
Would it be cleaner to consolidate `Set` and `SetNoCopyNoFinalize` into one API, e.g., something like
```
type SetOptions struct {
  CopyID bool
}

func (m *Map) Set(id ID, value ValueType, o SetOptions) {
  var opts mapSetOptions
  if o.CopyID {
    opts = mapSetOptions{
      copyKey: true,   
      finalizeKey: true,
    })
  }
  m.set(id, opts)
} 
```
?

Reply:
The thing I didn't like about that was needing to pass in SetOptions even when you don't want to specify anything special.